src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDesignTimeNodeWriter.cs

```diff
@@ -457,6 +457,27 @@ public override void WriteComponent(CodeRenderingContext context, ComponentInter
                 context.CodeWriter.Write(");");
                 context.CodeWriter.WriteLine();
             }
+
+            // We want to generate something that references the Component type to avoid
+            // the "usings directive is unnecessary" message.
+            // Looks like:
+            // __o = typeof(SomeNamespace.SomeComponent);
+            context.CodeWriter.Write(DesignTimeVariable);
+            context.CodeWriter.Write(" = ");
+            context.CodeWriter.Write("typeof(");
+            context.CodeWriter.Write(node.TagName);
+            if (node.Component.IsGenericTypedComponent())
+            {
+                context.CodeWriter.Write("<");
+                var typeArgumentCount = node.Component.GetTypeParameters().Count();
+                for (var i = 1; i < typeArgumentCount; i++)
+                {
+                    context.CodeWriter.Write(",");
```

Comment:
There is no need to worry about nested generics. All we need to know is the count of generic type arguments.
```
Counter => typeof(Counter)
Counter<string, int> => typeof(Counter<,>)
Counter<string, string, int> => typeof(Counter<,,>)
```

Reply:
I see 👍

Reply:
Would it be crazy to just avoid using qualified names during design time? 

So instead of `builder.OpenComponent<MyApp.FooBar.Counter>()` just generate `builder.OpenComponent<Counter>()`. 

One of the things we may need to do eventually is give a line-mapping to one of the statements we generate so that FAR can target it.

Reply:
> So instead of `builder.OpenComponent<MyApp.FooBar.Counter>()` just generate `builder.OpenComponent<Counter>()`

But we don't generate any `builder.OpenComponent<>()` during design time (Except for type inference for generic components). Is this a bug?

Here is what gets generated for this input,
```cshtml
@using MvcWithComponents.Components

<Counter></Counter>

<Generic Item="17"></Generic> 
```

<details><summary>Generated design time code:</summary>
<p>

```csharp
// <auto-generated/>
#pragma warning disable 1591
namespace MvcWithComponents
{
    #line hidden
    using TModel = global::System.Object;
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;
    using Microsoft.AspNetCore.Components;
#line 1 "C:\Users\ajbaaska\Source\Repos\MvcWithComponents\MvcWithComponents\TestComponent.razor"
using MvcWithComponents.Components;

#line default
#line hidden
    public class __TestComponent : Microsoft.AspNetCore.Components.ComponentBase
    {
        #pragma warning disable 219
        private void __RazorDirectiveTokenHelpers__() {
        }
        #pragma warning restore 219
        #pragma warning disable 0414
        private static System.Object __o = null;
        #pragma warning restore 0414
        #pragma warning disable 1998
        protected override void BuildRenderTree(Microsoft.AspNetCore.Components.RenderTree.RenderTreeBuilder builder)
        {
            base.BuildRenderTree(builder);
            builder.AddAttribute(-1, "ChildContent", (Microsoft.AspNetCore.Components.RenderFragment)((builder2) => {
            }
            ));
            __o = typeof(Counter);
            __Blazor.MvcWithComponents.__TestComponent.TypeInference.CreateGeneric_0(builder, -1, -1, 
#line 5 "C:\Users\ajbaaska\Source\Repos\MvcWithComponents\MvcWithComponents\TestComponent.razor"
               17

#line default
#line hidden
            );
            __o = typeof(Generic<>);
        }
        #pragma warning restore 1998
    }
}
namespace __Blazor.MvcWithComponents.__TestComponent
{
    #line hidden
    internal static class TypeInference
    {
        public static void CreateGeneric_0<TItem>(global::Microsoft.AspNetCore.Components.RenderTree.RenderTreeBuilder builder, int seq, int __seq0, TItem __arg0)
        {
        builder.OpenComponent<global::MvcWithComponents.Components.Generic<TItem>>(seq);
        builder.AddAttribute(__seq0, "Item", __arg0);
        builder.CloseComponent();
        }
    }
}
#pragma warning restore 1591

```

</p>
</details>

Reply:
@rynowak ping

Reply:
> But we don't generate any builder.OpenComponent<>() during design time (Except for type inference for generic components). Is this a bug?

It would be a good thing to add I think. It's better than doing something artificial IMO.

Reply:
> It would be a good thing to add I think. It's better than doing something artificial IMO.

I tried this but it gets sort of complicated because we need to do things differently when we do type inference vs when we don't. Generating just a `typeof()` in all cases seems cleaner and simpler.

Reply:
ok